### PR TITLE
Matching data types to internal PySpark and pandas' behavior.

### LINKED ISSUE #1840: Type hints do not works for many spark types
Type hints do not support float type and many others.
Sample code that does not work:
```
    import databricks.koalas as ks
    import pandas as pd

    df = spark_session.createDataFrame([
        (1, 1.,),
    ], 'a long, b float')
    kdf = df.to_koalas()

    def f(pdf: pd.DataFrame) -> ks.DataFrame[zip(kdf.columns, kdf.dtypes)]:
        return pdf

    print(kdf.koalas.apply_batch(f))
```

It throws error:
```
TypeError: Type <class 'numpy.float32'> was not understood.
from databricks/koalas/typedef/typehints.py:128
```

Changing `b` column type from `float` to `double` fixes this error
`strinng` columns are also not supported in this scenario

### PR DESCRIPTION
**Pull request by itholic**
This PR basically includes several fixes for matching data types to internal PySpark and pandas' behavior :
- Added `np.float32` and `"float32"` (matched to `FloatType`)
    ```python
    >>> ks.Series([10]).astype(np.float32)
    0    10.0
    dtype: float32

    >>> ks.Series([10]).astype("float32")
    0    10.0
    dtype: float32
    ```
- Added `np.datetime64` and `"datetime64[ns]"` (matched to `TimestampType`) 
    ```python
    >>> ks.Series(["2020-10-26"]).astype(np.datetime64)
    0   2020-10-26
    dtype: datetime64[ns]

    >>> ks.Series(["2020-10-26"]).astype("datetime64[ns]")
    0   2020-10-26
    dtype: datetime64[ns]
    ```
- Fixed `np.int` to match `LongType`, not `IntegerType`.
    ```python
    >>> pd.Series([100]).astype(np.int)
    0    100.0
    dtype: int64

    >>> ks.Series([100]).astype(np.int)
    0    100.0
    dtype: int32  # This fixed to `int64` now.
    ```
- Fixed `np.float` to match `DoubleType`, not `FloatType`.
    ```python
    >>> pd.Series([100]).astype(np.float)
    0    100.0
    dtype: float64

    >>> ks.Series([100]).astype(np.float)
    0    100.0
    dtype: float32  # This fixed to `float64` now.
    ```
- Added test, `test_typedef.py::TypeHintTests::test_as_spark_type`
- Added more tests for `test_series.py::SeriesTest::test_astype`
- Fixed related doc tests

Should resolve #1840

### CODE REVIEW

File: databricks/koalas/typedef/typehints.py [hunk 1]
```diff
@@ -93,41 +93,48 @@ class NameTypeHolder(object):
...
+        return types.IntegerType()
+    elif tpe in (int, np.int, np.int64, "int", "int64", "long", "bigint"):
+        return types.LongType()
+    elif tpe in (np.int16, "int16", "short"):
+        return types.ShortType()
+    # StringType
+    elif tpe in (str, "str", "string"):
+        return types.StringType()
+    # TimestampType
+    elif tpe in (datetime.datetime, np.datetime64, "datetime64[ns]"):
```
Comment (ueshin) [inline databricks/koalas/typedef/typehints.py:136]: also `pd.Timestamp`?

Comment (itholic) [reply databricks/koalas/typedef/typehints.py:136]: I think maybe then we have inconsistency behavior with pandas for `astype` ??

```python
>>> pd.Series(["2020-10-20"]).astype(pd.Timestamp)
Traceback (most recent call last):
...
TypeError: dtype '<class 'pandas._libs.tslibs.timestamps.Timestamp'>' not understood

>>> ks.Series(["2020-10-20"]).astype(pd.Timestamp)
0   2020-10-20
dtype: datetime64[ns]
```

What do you think about this ??

Comment (ueshin) [reply databricks/koalas/typedef/typehints.py:136]: Got it. Let's not add it for now. We might need to add it later anyway since the function is used not only in `astype`.

File: databricks/koalas/typedef/typehints.py [hunk 2]
```diff
@@ -93,41 +93,48 @@ class NameTypeHolder(object):
...
     """
-    if tpe in (str, "str", "string"):
-        return types.StringType()
-    elif tpe in (bytes,):
+    # ArrayType
+    if tpe in (np.ndarray,):
+        # TODO: support other child types
+        return types.ArrayType(types.StringType())
+    # BinaryType
+    elif tpe in (bytes, np.character):
```
Comment (ueshin) [inline databricks/koalas/typedef/typehints.py:109]: `np.bytes_`?

File: databricks/koalas/typedef/typehints.py [hunk 3]
```diff
@@ -93,41 +93,48 @@ class NameTypeHolder(object):
...
         return types.BinaryType()
+    # BooleanType
+    elif tpe in (bool, np.bool, "bool", "boolean"):
+        return types.BooleanType()
+    # DateType
+    elif tpe in (datetime.date,):
+        return types.DateType()
+    # NumericType
     elif tpe in (np.int8, "int8", "byte"):
```
Comment (ueshin) [inline databricks/koalas/typedef/typehints.py:118]: `np.byte`?

File: databricks/koalas/typedef/typehints.py [hunk 4]
```diff
@@ -93,41 +93,48 @@ class NameTypeHolder(object):
...
+    elif tpe in (float, np.float, np.float64, "float", "float64", "double"):
+        return types.DoubleType()
+    elif tpe in (np.float32, "float32"):
+        return types.FloatType()
+    elif tpe in (np.int32, "int32"):
+        return types.IntegerType()
+    elif tpe in (int, np.int, np.int64, "int", "int64", "long", "bigint"):
+        return types.LongType()
+    elif tpe in (np.int16, "int16", "short"):
+        return types.ShortType()
```
Comment (ueshin) [inline databricks/koalas/typedef/typehints.py:131]: Why did you change the order of `if-else`s?

File: databricks/koalas/typedef/typehints.py [hunk 5]
```diff
@@ -93,41 +93,48 @@ class NameTypeHolder(object):
...
+    # StringType
+    elif tpe in (str, "str", "string"):
```
Comment (ueshin) [inline databricks/koalas/typedef/typehints.py:133]: `np.unicode_`?

File: databricks/koalas/typedef/typehints.py (hunk 2 above)
Comment (ueshin) [reply databricks/koalas/typedef/typehints.py:109]: `np.string_`?

**Review by ueshin — COMMENTED** [review]
Also could you check whether pandas supports string notation like:

```
'?' | boolean
'b' | (signed) byte
'B' | unsigned byte
'i' | (signed) integer
'u' | unsigned integer
'f' | floating-point
'c' | complex-floating point
'm' | timedelta
'M' | datetime
'O' | (Python) objects
'S', 'a' | zero-terminated bytes (not recommended)
'U' | Unicode string
'V' | raw data (void)
```
https://numpy.org/doc/stable/reference/arrays.dtypes.html

File: databricks/koalas/typedef/typehints.py (hunk 4 above)
Comment (itholic) [reply databricks/koalas/typedef/typehints.py:131]: I re-order them according to the order of the class hierarchy described in the [spark documentation](https://spark.apache.org/docs/3.0.0-preview/api/java/org/apache/spark/sql/types/package-tree.html). If this is not necessary, will revert it.

**Comment by itholic** [comment]
Thanks, @ueshin . Added the missing types.

**Comment by itholic** [comment]
> Also could you check whether pandas supports string notation like:
> 
> ```
> '?' | boolean
> 'b' | (signed) byte
> 'B' | unsigned byte
> 'i' | (signed) integer
> 'u' | unsigned integer
> 'f' | floating-point
> 'c' | complex-floating point
> 'm' | timedelta
> 'M' | datetime
> 'O' | (Python) objects
> 'S', 'a' | zero-terminated bytes (not recommended)
> 'U' | Unicode string
> 'V' | raw data (void)
> ```
> 
> https://numpy.org/doc/stable/reference/arrays.dtypes.html

Yep, I'll check it !

File: databricks/koalas/typedef/typehints.py (hunk 4 above)
Comment (ueshin) [reply databricks/koalas/typedef/typehints.py:131]: ah, I see.

File: databricks/koalas/typedef/typehints.py [hunk 6]
```diff
@@ -93,41 +93,48 @@ class NameTypeHolder(object):
...
-    elif tpe in (int, "int", np.int, np.int32):
-        return types.IntegerType()
-    elif tpe in (np.int64, "int64", "long", "bigint"):
-        return types.LongType()
-    elif tpe in (float, "float", np.float):
-        return types.FloatType()
-    elif tpe in (np.float64, "float64", "double"):
-        return types.DoubleType()
     elif tpe in (decimal.Decimal,):
         return types.DecimalType(38, 18)
```
Comment (ueshin) [inline databricks/koalas/typedef/typehints.py:121]: Shall we leave a TODO comment to pick up the precision and the scale somehow?

File: databricks/koalas/typedef/typehints.py [hunk 7]
```diff
@@ -93,41 +93,49 @@ class NameTypeHolder(object):
...
+    elif tpe in (np.float32, "float32", "f"):
+        return types.FloatType()
+    elif tpe in (np.int32, "int32", "i"):
+        return types.IntegerType()
+    elif tpe in (int, np.int, np.int64, "int", "int64", "long", "bigint"):
+        return types.LongType()
+    elif tpe in (np.int16, "int16", "short"):
+        return types.ShortType()
+    # StringType
+    elif tpe in (str, np.unicode_, "str", "string"):
```
Comment (ueshin) [inline databricks/koalas/typedef/typehints.py:134]: maybe `'U'` is here: 

```py
>>> pd.Series([1,2,3]).astype('U')
0    1
1    2
2    3
dtype: object
>>> pd.Series([1,2,3]).astype('U')[0]
'1'

>>> np.dtype('U') == np.unicode_
True
```

File: databricks/koalas/typedef/typehints.py [hunk 8]
```diff
@@ -93,41 +93,49 @@ class NameTypeHolder(object):
...
-        return types.ShortType()
-    elif tpe in (int, "int", np.int, np.int32):
-        return types.IntegerType()
-    elif tpe in (np.int64, "int64", "long", "bigint"):
-        return types.LongType()
-    elif tpe in (float, "float", np.float):
-        return types.FloatType()
-    elif tpe in (np.float64, "float64", "double"):
-        return types.DoubleType()
+    # TODO: considering about the precision & scale for decimal type.
```
Comment (ueshin) [inline databricks/koalas/typedef/typehints.py:120]: Shall we move this into `elif tpe in (decimal.Decimal,):` block?

File: databricks/koalas/tests/test_series.py [hunk 9]
```diff
@@ -1196,7 +1196,23 @@ def test_astype(self):
...
+        self.assert_eq(kser.astype("int16"), pser.astype("int16"))
+        self.assert_eq(kser.astype("int32"), pser.astype("int32"))
+        self.assert_eq(kser.astype("int64"), pser.astype("int64"))
+        self.assert_eq(kser.astype(np.float), pser.astype(np.float))
+        self.assert_eq(kser.astype(np.float32), pser.astype(np.float32))
+        self.assert_eq(kser.astype(np.float64), pser.astype(np.float64))
+        self.assert_eq(kser.astype("float"), pser.astype("float"))
+        self.assert_eq(kser.astype("float32"), pser.astype("float32"))
+        self.assert_eq(kser.astype("float64"), pser.astype("float64"))
         self.assert_eq(kser.astype(bool), pser.astype(bool))
```
Comment (ueshin) [inline databricks/koalas/tests/test_series.py:1216]: Shall we add more tests for the string notation you added later?

**Review by ueshin — COMMENTED** [review]
Otherwise, LGTM.

File: databricks/koalas/tests/test_series.py (hunk 9 above)
Comment (itholic) [reply databricks/koalas/tests/test_series.py:1216]: Sure, I'll add them.
Thanks!

File: databricks/koalas/typedef/typehints.py [hunk 10]
```diff
@@ -93,41 +93,50 @@ class NameTypeHolder(object):
...
     - the built-in types in numpy
     - list of pairs of (field_name, type)
     - dictionaries of field_name -> type
-    - python3's typing system
+    - Python3's typing system
     """
-    if tpe in (str, "str", "string"):
-        return types.StringType()
-    elif tpe in (bytes,):
+    # TODO: Add "boolean" and "string" types.
```
Comment (itholic) [inline databricks/koalas/typedef/typehints.py:104]: I added "boolean" and "string" as TODO item since they work differently from "bool" and "str" respectively.

First, "boolean" casts the type from "bool-like values" to `boolean` whereas "bool" casts the type from various types to `bool`.

```python
>>> pd.Series([1, 2, 3]).astype("boolean")
Traceback (most recent call last):
...
TypeError: Need to pass bool-like values

>>> pd.Series([True, False, True]).astype("boolean")
0     True
1    False
2     True
dtype: boolean
```

Next, "string" casts the type from "sequence of string or pandas.NA" to `string` whereas "str" casts the type from various types to `object `.

```python
>>> pd.Series([1, 2, 3]).astype("string")
Traceback (most recent call last):
...
ValueError: StringArray requires a sequence of strings or pandas.NA

>>> pd.Series(['1', '2', '3']).astype("string")
0    1
1    2
2    3
dtype: string
```

In short, we need to add the new type `boolean` and `string` to match these behaviors.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by itholic** [comment]
I think It's good to go for now.

How about let's discuss the remaining things in the separated PR ?

**Review by ueshin — APPROVED** [review]
LGTM.

**Comment by ueshin** [comment]
Thanks! I'd merge this now.
@itholic Could you move onto the the remaining topics in the separate PRs?

**Comment by itholic** [comment]
Thanks, @ueshin !
Sure I'll open a PR soon.

**Comment by nanotellez** [comment]
I just made sure I had the latest koalas version (1.5.0) and tested this and it is not working as described above. 
```
ks.Series([10]).astype(np.float32)
```
Produced the following error:
```

ks.Series([10]).astype(np.float32)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-10-4576f10d772a> in <module>
----> 1 ks.Series([10]).astype(np.float32)

~/opt/anaconda3/envs/pipelines_env/lib/python3.7/site-packages/databricks/koalas/series.py in astype(self, dtype)
   1002         from databricks.koalas.typedef import as_spark_type
   1003 
-> 1004         spark_type = as_spark_type(dtype)
   1005         if not spark_type:
   1006             raise ValueError("Type {} not understood".format(dtype))

~/opt/anaconda3/envs/pipelines_env/lib/python3.7/site-packages/databricks/koalas/typedef/typehints.py in as_spark_type(tpe)
    123         return types.ArrayType(types.StringType())
    124     else:
--> 125         raise TypeError("Type %s was not understood." % tpe)
    126 
    127 

TypeError: Type <class 'numpy.float32'> was not understood.
```

**Comment by itholic** [comment]
@nanotellez 

Hmm... I couldn't reproduce the same error.

Could you let me know how did you install Koalas ??

I installed simply via `pip install koalas`, and tested, it worked fine.

```python
>>> ks.__version__
'1.5.0'

>>> ks.Series([10]).astype(np.float32)
0    10.0
dtype: float32
```

**Comment by nanotellez** [comment]
Well... upon checking again, I realized that the version I had was in fact *not* the latest one. The update I ran must not have gone through for some reason, and I did not realize it. Now that I carefully ran the update, it is working exactly as intended. My apologies!

**Comment by itholic** [comment]
@nanotellez No problem. Glad to hear that you fixed !! :D
